src/System.Management.Automation/engine/parser/DynamicKeyword.cs

```diff
@@ -0,0 +1,2759 @@
+
+            return GlobalKeywords.Remove(name);
+        }
+
+        /// <summary>
+        /// Add a dynamic keyword to the global namespace, so that it is defined everywhere
+        /// </summary>
+        /// <param name="keywordToAdd">the keyword to add</param>
+        /// <returns>true if an older keyword was overwritten, false otherwise</returns>
+        public bool AddGlobalDynamicKeyword(DynamicKeyword keywordToAdd)
+        {
+            if (keywordToAdd == null)
+            {
+                throw new PSArgumentException(nameof(keywordToAdd));
+            }
+
+            string name = keywordToAdd.Keyword;
+            if (String.IsNullOrEmpty(name))
+            {
+                throw new PSArgumentException(nameof(keywordToAdd.Keyword));
+            }
+
+            bool result = GlobalKeywords.Remove(name);
+            GlobalKeywords.Add(name, keywordToAdd);
+            return result;
+        }
+
+        /// <summary>
+        /// Check whether a keyword by the given name is globally defined
+        /// </summary>
+        /// <param name="name">the name of the keyword to check</param>
+        /// <returns>true if a keyword with the given name exists in the global namespace, false otherwise</returns>
+        public bool IsGloballyDefined(string name)
+        {
+            return GlobalKeywords.ContainsKey(name);
+        }
+
+        /// <summary>
+        /// Get a keyword that is defined inside any of the current enclosing scopes by name,
+        /// including the global namespace, or null if no such keyword exists
+        /// </summary>
+        /// <param name="name">the name of the keyword to search for</param>
+        /// <returns>the keyword of the given name, or null if no such keyword exists</returns>
+        public DynamicKeyword GetScopedDynamicKeyword(string name)
+        {
+            DynamicKeyword keyword;
+
+            // First search upward through the enclosing scopes
+            foreach (var enclosingKeyword in DynamicKeywordScope)
+            {
+                enclosingKeyword.InnerKeywords.TryGetValue(name, out keyword);
+                if (keyword != null)
+                {
+                    return keyword;
+                }
+            }
+
+            // Then search the global namespace
+            keyword = GetGlobalDynamicKeyword(name);
+            if (keyword != null)
+            {
+                return keyword;
+            }
+
+            // Admit defeat
+            return null;
+        }
+
+        /// <summary>
+        /// Check whether a keyword of the given name is defined in any of the
+        /// current enclosing scopes, including the global namespace
+        /// </summary>
+        /// <param name="name">the name of the keyword to search for</param>
+        /// <returns>true if the keyword exists in an enclosing scope or the global namespace, otherwise false</returns>
+        public bool IsKeywordDefinedInCurrentScope(string name)
```

Comment:
I think I named the API poorly -- I had "current scope" in mind as the total scope we're currently in, not the most local scope. But it would be better as maybe `IsKeywordInScope`.